### ISSUE TITLE
Add SignCheck feature to exclude a nupkg from validation if it contains a marker file

### ISSUE THREAD

**Issue by dagood** [dotnet/arcade#6810]
- [ ] This issue is blocking 
- [ ] This issue is causing unreasonable pain

This is blocking arcade-powered source-build until either the feature is implemented, or we decide we won't do it and instead go with an alternate approach like excluding the nupkg from validation using a wildcard in the yaml (https://github.com/dotnet/arcade/issues/6806).

---

The arcade-powered source-build plan involves every repo making an intermediate nupkg. They look like:
`Microsoft.SourceBuild.Intermediate.{repo-name}.{version}.nupkg`
they are built on Linux and unsigned, with a source-built unsigned version of every nupkg inside in `artifacts/`.

I asked @chcosta about how to exclude this from signing validation by default from the Arcade side, and I liked this idea the best:

Add a marker file to the intermediate nuget package that SignCheck would obey. If SignCheck sees the marker at the root of a nupkg, it allows the nupkg and all its contents (recursive) to be unsigned.

This lets us leave SignCheck command lines and the yaml exactly the same. The marker could be an empty `.intentionally-not-signed` file at the nupkg root, or something weirder or more complicated if there's concern about this being used inappropriately. Here's a sample intermediate nupkg with nested nupkg to maybe try it out on later: [Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.zip](https://github.com/dotnet/arcade/files/5822275/Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.zip)

---

There are some other alternatives that I think aren't quite as good, but some don't require SignCheck changes:

* The `eng/SignCheckExclusionsFile.txt` is checked into each repo, and I don't think we should require every repo to have that file just because it participates in source-build.
* We could edit the `eng/common` template yaml to use an `itemPattern` that avoids downloading the intermediate nupkgs by filename convention.
  * The filename prefix seems like it could only ever apply to intermediate nupkgs, but there's still a degree of uncertainty.
  * Makes it harder to repro locally: have to manually filter out PackageArtifacts yourself before running SignCheck.
* Add `ExcludeFiles` option to SignCheck that accepts wildcards, like InputFiles
  * This is similar to the yaml itemPattern, except it requires a signcheck change.  The benefit being that the filtering is a feature of the tool, rather than having to code around the problem in yaml
  * (@chcosta raised this as a possibility. Same issue where a wildcard doesn't quite seem ideal. However, `ExcludeFiles` would be included in the SignCheck command, so it wouldn't be as bad as a yaml change for local reproability.)

---

@chcosta @mmitche @joeloff is the marker a reasonable feature to add?

**Comment by mmitche** [comment]
I don't love the idea. It doesn't have broad applicability across artifact types (e.g. pkg or deb) and feels out of pattern with the rest of the things arcade does around signing. What if....

Today we have Sign.props in arcade which applies defaults for repos, and then repos apply their own info on top of that with eng /Signing.props. What if we did a similar pattern for exclusions? The arcade defaults would include the intermediate nupkgs somehow.

**Comment by joeloff** [comment]
@dagood just to be clear, when you want to exclude the nupkg, do you care about the content at all? There's two levels of signing here, one for the contents (it anything is signable) and one for the package itself.

**Comment by dagood** [comment]
> Today we have Sign.props in arcade which applies defaults for repos, and then repos apply their own info on top of that with eng /Signing.props. What if we did a similar pattern for exclusions? The arcade defaults would include the intermediate nupkgs somehow.

That sounds fine to me. I don't see how to get that to work with the exclusion patterns we have right now, maybe I'm missing something or we need some new exclusion syntax. I tried adding to these existing built-in exclusions:

https://github.com/dotnet/arcade/blob/86fde30efc453597aa7be8af2d92569c38b75476/src/SignCheck/SignCheck/SignCheck.cs#L173-L174

```cs
            Exclusions.Add(new Exclusion("Microsoft.SourceBuild.Intermediate.*.nupkg;;Arcade-powered source-build intermediate nupkg"));
            Exclusions.Add(new Exclusion("*;Microsoft.SourceBuild.Intermediate.*.nupkg;Arcade-powered source-build intermediate nupkg"));
            Exclusions.Add(new Exclusion("*;Microsoft.SourceBuild.Intermediate.*.nupkg*;Arcade-powered source-build intermediate nupkg"));
```

and that only got me the top two levels, not the DLLs inside:

```
[File] Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.nupkg, Signed: False, Excluded
  [File] 110b2d7e344e0481efb98ab3ddfe008b.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.Build.Tasks.Git.1.1.0-ci.nupkg
    [File] 6f7cc6785913a2495ed1a216c179d481.dll, Signed: False, Full Name: tools/net472/cs/Microsoft.Build.Tasks.Git.resources.dll [Error] HRESULT: 800b0100 (No signature was present in the subject)
...
```

> when you want to exclude the nupkg, do you care about the content at all? There's two levels of signing here, one for the contents (it anything is signable) and one for the package itself.

The nupkg and its contents (and its content's contents) are all intentionally not signed, and I want to exclude all of it.

**Comment by joeloff** [comment]
Where can I get a copy of a sample package?

**Comment by dagood** [comment]
I'm using this one:

> Here's a sample intermediate nupkg with nested nupkg to maybe try it out on later: [Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.zip](https://github.com/dotnet/arcade/files/5822275/Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.zip)

With `-i "C:\Users\dagood\Desktop\Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.nupkg" -r`

**Comment by joeloff** [comment]
I think I see what's going on. The nupkg contains other nupkgs. Exclusions are always associated with a container/parent.

This excludes the nupkg and the top-level files. But it won't exclude the files inside the nested nupkgs

```
Microsoft.SourceBuild*.nupkg;;
*;Microsoft.SourceBuild*.nupkg;
```

However, if you did something like this:
```
Microsoft.SourceBuild*.nupkg;;
*;Microsoft.SourceBuild*.nupkg;
*;110b2d7e344e0481efb98ab3ddfe008b.nupkg;
```

It will exclude a lot more. The idea was always to avoid over exclusions and potentially miss something. However, this is a bug because the parent container should be match on the relative container path, e.g.

```
*;artifacts/Shipping/Microsoft.Build.Tasks.Git.1.1.0-ci.nupkg;
```

**Comment by dagood** [comment]
I'm not sure what you're suggesting, something like this, once a bug is fixed where the temp name is used instead of the relative path?
```
Microsoft.SourceBuild*.nupkg;;
*;Microsoft.SourceBuild*.nupkg;
*;artifacts/*.nupkg;
```
(To make false excludes less likely, modify the intermediate nupkg infra to put things in `artifacts/source-built/` maybe.)

**Comment by joeloff** [comment]
Ok, I just ran with a recent copy of the tool and this works:

```
Microsoft.SourceBuild*.nupkg;;
*;Microsoft.SourceBuild*.nupkg;
*;*.nupkg;
```

```
[File] Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.nupkg, Signed: False, Excluded
  [File] 110b2d7e344e0481efb98ab3ddfe008b.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.Build.Tasks.Git.1.1.0-ci.nupkg
  [File] f37bcebbcbe1b271fae816d94a455513.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.AzureDevOpsServer.Git.1.1.0-ci.nupkg
  [File] 35bf5bb8eba09da0cf4130bf26bc8a69.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.AzureRepos.Git.1.1.0-ci.nupkg
  [File] a31b612bc424be7d1edfa65f44e7fb32.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.Bitbucket.Git.1.1.0-ci.nupkg
  [File] cddb5d4093cfc5252f33f850b2095bd8.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.Common.1.1.0-ci.nupkg
  [File] 9b0a5d7f39879cb61224667f47799faf.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.Gitea.1.1.0-ci.nupkg
  [File] dd658da62b7975361c387be58ecf35f1.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.GitHub.1.1.0-ci.nupkg
  [File] 0a4aded6ef9770c8c522d21bb9906e71.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.GitLab.1.1.0-ci.nupkg
  [File] e115dbc75a915d84a7a3aea4f1bda2ad.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.GitWeb.1.1.0-ci.nupkg
  [File] 791cc9bc4caba857c81979ddf6f6524f.nupkg, Signed: False, Excluded, Full Name: artifacts/Shipping/Microsoft.SourceLink.Tools.1.1.0-ci.nupkg

No signing issues found
Total Time: 00:00:04.9655249
Total Files: 362, Signed: 0, Unsigned: 0, Skipped: 91, Excluded: 362, Skipped & Excluded: 91
```

**Comment by dagood** [comment]
Ah, I see. Same result here. Would it be reasonable for me to add these lines to the default exclusions right now, to get source-build unblocked?

```cs
Exclusions.Add(new Exclusion("Microsoft.SourceBuild.Intermediate.*.nupkg;;Arcade-powered source-build intermediate nupkg"));
Exclusions.Add(new Exclusion("*;Microsoft.SourceBuild.Intermediate.*.nupkg;Arcade-powered source-build intermediate nupkg"));
Exclusions.Add(new Exclusion("*;*.nupkg;Arcade-powered source-build intermediate nupkg"));
```

I suppose it's not common to have a nupkg inside a nupkg, so the generality is maybe not too bad. 😄 Then once the relative path bug is fixed, we could make the exclusion more specific.

**Comment by joeloff** [comment]
Yeah, we could. The ideal fix would be to add multiple exclusion files (I don't think it supports that right now). That way each repo can add customizations with Arcade supplying a base exclusion set maybe.

There's also another option (this was ASP.NET originally). If the comment contains "DO-NOT-SIGN" it will be treated as an exclusion as well. It was a few years back, so I can't recall the exact use case

**Comment by dagood** [comment]
Agreed, first thing I looked for was a way to plug in more exclusion files, but didn't see any. Took me a bit of poking around to actually see that there were some defaults defined in the codebase. 😄

It would be nice to only include these exclusions in repos that are actually in arcade-powered source-build, to reduce possibility of bad excludes. There will be some repos that will probably never get in source-build. (Anything tied to WindowsDesktop, basically.)

> If the comment contains "DO-NOT-SIGN" it will be treated as an exclusion as well.

Ah yeah, dotnet/runtime uses this to ensure the apphost templates aren't signed.

It's not very important to source-build to make sure these files *aren't* signed, but might as well include that to save people some signing time/flakiness if things get signed unintentionally.

**Comment by dagood** [comment]
Good catch in https://github.com/dotnet/arcade/pull/6812 --this pattern excludes all files in *any* nupkg, even the non-source-build ones that should be signed:

```
*;*.nupkg;
```

And even if the relative path bug were fixed, there's no way to get the current pattern implementation to work with the intermediate packages as they are now. This pattern would apply to every file in the "real" copy of `Microsoft.SourceLink.Gitea.1.1.0-ci.nupkg` as well as the source-built copy:

```
*;Microsoft.SourceLink.Gitea.1.1.0-ci.nupkg;
```

The source-built and real packages are built in the same official build, so we can't toggle the rule on and off as a fix. It seems like we still need some way to match the parent's parent's name.

**Comment by dagood** [comment]
I'm starting to think we should go with the simple/known solution, accepting the downsides for now, and revisit a more integrated generalized exclusion later. (Simple solution: add a DownloadBuildArtifacts exclusion in yaml so we never even pass this file to SignCheck: https://github.com/dotnet/arcade/issues/6806)

**Comment by dagood** [comment]
@joeloff did you make any headway in adding exclusion support for these intermediate nupkgs?

**Comment by joeloff** [comment]
Yes and no. The exclusion processing will need a bit of a redesign to better handle this, but it should be feasible.

**Comment by joeloff** [comment]
I've got the tool to generate "virtual paths", e.g. 

```Microsoft.SourceBuild.Intermediate.sourcelink.1.1.0-ci.nupkg/artifacts/Shipping/Microsoft.SourceLink.GitLab.1.1.0-ci.nupkg/tools/net5.0/cs/Microsoft.SourceLink.GitLab.resources.dll```

I'm looking to see if we can leverage the standard file exclusion to act on the virtual path too, for example, something like

```*/artifacts/*.nupkg/*.dll``` should then potentially exclude DLLs in nested nupkgs that are leafs in the archive hierarchy I think

**Comment by joeloff** [comment]
With an exclusion like ```*/artifacts/*/Microsoft.SourceLink.GitLab.1.1.0-ci.nupkg/*.dll;;``` I can now see the following being reported:

```
[File] 0a4aded6ef9770c8c522d21bb9906e71.nupkg, Signed: False, Full Name: artifacts/Shipping/Microsoft.SourceLink.GitLab.1.1.0-ci.nupkg
    [File] 4e82e7f3ac71cbe89e40763ec5b4696c.dll, Signed: False, Excluded, Full Name: tools/net5.0/Microsoft.SourceLink.GitLab.dll
    [File] 5a83285ef569f48c7b2875e95d19b453.dll, Signed: False, Excluded, Full Name: tools/net472/Microsoft.SourceLink.GitLab.dll
```

**Comment by dagood** [comment]
Done, by https://github.com/dotnet/arcade/pull/6822. Thanks!

I've opened https://github.com/dotnet/arcade/issues/6922 to follow up and use this feature to remove the YAML-side workaround.
